modules/core/src/main/scala/fs2/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -364,20 +337,16 @@ private[kafka] final class KafkaConsumerActor[F[_], K, V](
   private[this] val offsetCommit: Map[TopicPartition, OffsetAndMetadata] => F[Unit] =
     offsets => {
       val commit =
-        Deferred[F, Either[Throwable, Unit]].flatMap { deferred =>
-          requests.enqueue1(Request.Commit(offsets, deferred)) >>
-            F.race(timer.sleep(settings.commitTimeout), deferred.get.rethrow)
-              .flatMap {
-                case Right(_) => F.unit
-                case Left(_) =>
-                  F.raiseError[Unit] {
-                    CommitTimeoutException(
-                      settings.commitTimeout,
-                      offsets
-                    )
-                  }
-              }
-        }
+        F.asyncF[Unit] { cb =>
+            requests.enqueue1(Request.Commit(offsets, cb))
+          }
+          .guarantee(context.shift)
+          .timeoutTo(settings.commitTimeout, F.raiseError[Unit] {
+            CommitTimeoutException(
+              settings.commitTimeout,
+              offsets
+            )
+          })
 
```

Comment:
Is this correct? Is the `.guarantee(context.shift)` needed, and if so is it in the right place? (@SystemFw )

Reply:
ah, that's interesting because you have a "split" async basically. You can put it here, or after the call to `commitAsync` (assuming that's the only place the callback is used non trivially)